documentation/ha.markdown

```diff
@@ -0,0 +1,263 @@
+
+        server_urls: ["https://primary-puppetdb.mycorp.net:8081", "https://replica-puppetdb.mycorp.net:8081"]
+        sticky_read_failover: true
+        command_broadcast: true
+
+   It is important that your primary PuppetDB appear first in this list, so it
+   will always be tried first when submitting new data from the Puppet master.
+
+4. Restart your PuppetDBs and your Puppet master.
+
+Deployment tips
+----
+
+- Each PuppetDB server should be configured to use a separate PostgreSQL
+  instance.
+
+- If you are using multiple compile masters, be sure that their clocks are
+  relatively in sync. (using NTP, for example) PuppetDB relies on timestamps for
+  discerning which data is newer for any given node, so the clocks on your
+  compile masters need to be within `runinterval` of each other.
+
+FAQ
+-----
+
+Q: *Why not just use PostgreSQL streaming replication?*
+
+A: The principal goal of PuppetDB's HA solution is to ensure that Puppet runs
+never fail, even in the face of losing machines and intermittent network
+connections. These precludes any possibility of coordinating writes among the
+members of the cluster; that is, this is a system that cares first about
+availability and second about consistency.
+
+Relational database replication systems tend to emphasize consistency. Those
+that are availability-focused must concern themselves with conflict resolution.
+For PuppetDB in particular, conflict resolution at the database level would be
+very complex and potentially intractable. Conversely, at the application level
+we can view the entities in very course terms (e.g. a whole catalog vs. the many
+database rows required to store its resources). This allows us to resolve any
+conflicts very predictably.
+
+If you need a read-only replica of your database for performance reasons, then
+PostgreSQL streaming replication works very well.
+
+Q: <a name="exportedResources"></a>*What's the deal with exported resources?*
+
+A: Exported resources are a great feature, but you need to be careful when using
+them. In particular: because PuppetDB is *eventually* consistent, changes to
+exported resources will *eventually* be visible on other nodes. (not
+immediately) This is true in any PuppetDB installation, but there is an added
+subtlety which you should be aware of when using an HA configuration:
+
+In failover scenarios, exported resources may appear to go back in time. This
+can happen if commands have been written to your primary PuppetDB but haven't yet
+been copied to the replica. If the primary PuppetDB becomes unreachable
+in this state, exported resource queries will be redirected to the replica,
+which does not yet have the latest data.
+
+This problem is significantly mitigated by using the command_broadcast flag in
+puppetdb.conf, as recommended. By pushing all data to the replica PuppetDB when
+doing the initial write, the replica should nearly always be up to date. But
```

Comment:
the replica should be nearly always up to date, but some failure cases still exist (for instance, X, Y, and Z). Because of this...